### PR TITLE
Fix errors on updating the draft and pending status

### PR DESCRIPTION
**Pull request by ingeniumed**
## Description

- Fix an issue introduced with #59 that stopped updates being possible to the draft and pending review statuses, as they were marked as restricted entirely instead of just to their names and slug.
- As part of this, I've also enabled updates to both the draft and pending review status's names. Internally, we reset the slug if it's ever updated so we never need to re-assign posts, etc. We also ensure that the new names are registered correctly.
- I have also updated the validation logic to be a lot more stricter as previously there were some edge cases that could get through.

Deletion of restricted statuses is still not allowed but this would be a good middle ground imo.

## Steps to Test

- Test from scratch that the names of draft and pending review can be changed
- Test also that new statuses with the same name as one existing or a restricted status cannot be made

### CODE REVIEW

**Comment by ingeniumed** [comment]
image

Draft and pending statuses can now be renamed with this PR!

File: modules/custom-status/custom-status.php [hunk 1]
```diff
@@ -583,24 +596,23 @@ public static function update_custom_status( int $status_id, array $args = [] ):
...
-		if ( self::is_restricted_status( $old_status->slug )
-		&& (
-			( isset( $args['name'] ) && $args['name'] !== $old_status->name )
-			||
-			( isset( $args['slug'] ) && $args['slug'] !== $old_status->slug )
-		) ) {
-			// translators: %s: Post status, like "Draft"
-			return new WP_Error( 'restricted', sprintf( __( 'Changing the name and slug of a restricted status (%s) is not allowed.', 'vip-workflow' ), $old_status->name ) );
+		// If the name was changed, we need to change the slug unless its banned from slug updates
+		if ( isset( $args['name'] ) && $args['name'] !== $old_status->name && ! self::is_status_banned_from_slug_changes( $old_status->slug ) ) {
```
Comment (alecgeatches) [inline modules/custom-status/custom-status.php:600]: I think we can remove the `! self::is_status_banned_from_slug_changes( $old_status->slug )` check from here, as the code unsets `$args['slug']` if the status is banned below anyway.

File: modules/custom-status/custom-status.php [hunk 2]
```diff
@@ -669,7 +681,7 @@ public static function delete_custom_status( int $status_id ): bool|WP_Error {
 
 		$old_status_slug = $old_status->slug;
 
-		if ( self::is_restricted_status( $old_status_slug ) ) {
+		if ( self::is_restricted_status( $old_status_slug ) && ! self::is_status_banned_from_slug_changes( $old_status_slug ) ) {
```
Comment (alecgeatches) [inline modules/custom-status/custom-status.php]: This code didn't seem to work for me - I can delete Draft and Pending Review statuses right now, in fact all statuses can be deleted until the last throws an error:

delete-all-statuses

Maybe this is supposed to be:

```suggestion
		if ( self::is_restricted_status( $old_status_slug ) || self::is_status_banned_from_slug_changes( $old_status_slug ) ) {
```

Comment (ingeniumed) [reply modules/custom-status/custom-status.php]: Yip, accidentally flipped the signs. I tested everything, and then did a bit of cleanup and a bad copy paste went in

File: modules/custom-status/custom-status.php (hunk 1 above)
Comment (ingeniumed) [reply modules/custom-status/custom-status.php:600]: Tbh, I put it there for safety reasons just in case
